File: code/backend/src/llm_geometry/lex/train.py
"""Training the Lexicon Lab's model — the recipe from `specs/006-.../architecture.md`.

The recipe is the source project's, with two **deliberate, documented** departures (both
marked `[DIFFERS]` in the architecture contract and stated in the UI):

1. **Weight decay applies only to the 2-D weight matrices** — `qkv_w`, `proj_w`, `fc1_w`,
   `fc2_w`, and `head_w` when untied. The source decays *every* parameter, including
   LayerNorm gains, biases, and both embedding tables; that is a known anti-pattern, and
   `decay_param_groups` makes the split inspectable rather than implicit.
2. **beta1 is fixed at 0.9.** The source's `OneCycleLR` also cycles beta1 between 0.95 and
   0.85; mirroring a second schedule across two languages buys nothing pedagogically.

Everything else is faithful: AdamW (0.9, 0.999, eps 1e-8), a one-cycle LR with `lr` as the
*peak* (so the first step runs at `lr/25`), global grad-norm clipping at 1.0, a contiguous
95/5 split of the token stream, and batches of start offsets drawn uniformly **with
replacement** — no epochs, no permutation, no document boundaries.

A run is seeded and reproducible on one platform. Bit-equality with the browser engine is
**not** claimed (platform BLAS, non-portable RNG streams) and the UI says so.

Non-finite loss raises `ComputeError` immediately. Training that has diverged must never
return a model and a number that look like a result.
"""

from __future__ import annotations

import math
import time
from contextlib import contextmanager
from dataclasses import dataclass, field
from typing import Callable, Iterator, Sequence

import numpy as np
import torch

from ..errors import ComputeError, InvalidParamError
from .config import (
    DEFAULT_BATCH,
    DEFAULT_LR,
    DEFAULT_SEED,
    DEFAULT_STEPS,
    DEFAULT_WEIGHT_DECAY,
    EOS_ID,
    GRAD_CLIP_NORM,
    MAX_STEPS,
    ONECYCLE_DIV_FACTOR,
    ONECYCLE_FINAL_DIV_FACTOR,
    ONECYCLE_PCT_START,
    PAD_ID,
    VAL_FRACTION,
)
from .model import LexConfig, LexModel
from .vocab import LexVocab, tokenize

#: The only parameters weight decay touches. Named explicitly rather than inferred from
#: `ndim == 2`, because `embed` and `pos` are 2-D too and must NOT be decayed.
DECAY_WEIGHTS = ("qkv_w", "proj_w", "fc1_w", "fc2_w", "head_w")

ProgressCb = Callable[["TrainProgress"], None]


@dataclass(frozen=True)
class TrainProgress:
    """One SSE tick. `lr` is the schedule's value for the step just taken."""

    step: int  # 1-based, so `step == total_steps` is the last tick
    total_steps: int
    loss: float
    lr: float
    elapsed_s: float

    def as_dict(self) -> dict[str, float | int]:
        return {
            "step": self.step,
            "total_steps": self.total_steps,
            "loss": self.loss,
            "lr": self.lr,
            "elapsed_s": self.elapsed_s,
        }


@dataclass
class TrainResult:
    """A finished run: the model, its trajectory, and the numbers the UI quotes."""

    model: LexModel
    history: list[dict[str, float | int]] = field(default_factory=list)
    first_loss: float = 0.0
    final_loss: float = 0.0
    val_loss: float = 0.0
    steps: int = 0
    seed: int = 0
    elapsed_s: float = 0.0

    def as_dict(self) -> dict[str, object]:
        return {
            "history": self.history,
            "first_loss": self.first_loss,
            "final_loss": self.final_loss,
            "val_loss": self.val_loss,
            "steps": self.steps,
            "seed": self.seed,
            "elapsed_s": self.elapsed_s,
            "config": self.model.cfg.as_dict(),
        }


# -- data ------------------------------------------------------------------------------


def token_stream(text: str, vocab: LexVocab) -> np.ndarray:
    """Encode a corpus as one id stream, with `<eos>` closing every non-blank line.

    Nursery rhymes are line-shaped: the line, not the paragraph, is the unit the model can
    plausibly learn to finish, so `<eos>` marks line ends. (The source emitted a separate
    `<nl>` token per line and `<eos>` per blank line; we have no `<nl>` row, and spending
    one of a 40-word budget's rows on one would be a real cost.)
    """
    ids: list[int] = []
    for line in text.splitlines():
        line_ids = vocab.encode(tokenize(line))
        if line_ids:
            ids.extend(line_ids)
            ids.append(EOS_ID)
    return np.asarray(ids, dtype=np.int64)


def split_stream(
    stream: np.ndarray, ctx: int, val_fraction: float = VAL_FRACTION
) -> tuple[np.ndarray, np.ndarray]:
    """Contiguous 95/5 train/val split — a prefix and a suffix, never shuffled.

    Both halves must hold at least one `ctx + 1` window, otherwise there is nothing to
    train or evaluate on and the caller is told exactly how short the corpus is.
    """
    stream = np.asarray(stream, dtype=np.int64)
    n_train = int((1.0 - val_fraction) * len(stream))
    train, val = stream[:n_train], stream[n_train:]
    span = ctx + 1
    if len(train) < span or len(val) < span:
        needed = int(math.ceil(span / min(1.0 - val_fraction, val_fraction)))
        raise InvalidParamError(
            f"corpus too short for ctx={ctx}: the {1 - val_fraction:.0%}/{val_fraction:.0%} "
            f"split gives {len(train)} training and {len(val)} validation tokens, and each "
            f"side needs at least {span}. Supply about {needed} tokens, or lower ctx."
        )
    return train, val


def sample_batch(
    split: np.ndarray, ctx: int, batch_size: int, gen: torch.Generator
) -> tuple[torch.Tensor, torch.Tensor]:
    """`batch_size` windows of `ctx + 1` tokens, start offsets uniform WITH replacement.

    Returns `(inputs, targets)`, both `(B, ctx)` — the classic shift-by-one pair, so every
    one of the `ctx` positions contributes to the loss. Because attention is causal, this
    is exactly `logits[:, :-1]` against `window[:, 1:]` on the full `ctx + 1` window.
    """
    high = len(split) - ctx  # last valid start: a full ctx+1 window must fit
    if high < 1:
        raise InvalidParamError(
            f"split of {len(split)} tokens cannot hold a {ctx + 1}-token training window"
        )
    starts = torch.randint(0, high, (batch_size,), generator=gen)
    data = torch.from_numpy(np.asarray(split, dtype=np.int64))
    windows = torch.stack([data[s : s + ctx + 1] for s in starts.tolist()])
    return windows[:, :-1].contiguous(), windows[:, 1:].contiguous()


def iter_eval_windows(split: np.ndarray, ctx: int, batch_size: int) -> Iterator[torch.Tensor]:
    """Non-overlapping `ctx + 1` windows, in order — a FIXED validation set.

    The source redrew ten random validation windows at every eval, so its `val` curve
    carried sampling noise and was not comparable point to point. This one is.
    """
    data = torch.from_numpy(np.asarray(split, dtype=np.int64))
    span = ctx + 1
    starts = list(range(0, len(data) - span + 1, ctx))
    for i in range(0, len(starts), batch_size):
        chunk = starts[i : i + batch_size]
        yield torch.stack([data[s : s + span] for s in chunk])


# -- the recipe ------------------------------------------------------------------------


def onecycle_lr(
    step: int,
    total_steps: int,
    peak_lr: float,
    pct_start: float = ONECYCLE_PCT_START,
    div_factor: float = ONECYCLE_DIV_FACTOR,
    final_div_factor: float = ONECYCLE_FINAL_DIV_FACTOR,
) -> float:
    """The one-cycle schedule, in closed form. `step` is 0-indexed; `peak_lr` is the peak.

    ``initial = peak/25``, ``final = initial/1e4``, cosine on both phases, warmup over
    ``w = round(pct_start · S)`` steps:

        i < w:  p = i/w,       lr = initial + (peak − initial)·(1 − cos(πp))/2
        i ≥ w:  p = (i−w)/(S−w), lr = final + (peak − final)·(1 + cos(πp))/2
    """
    if total_steps < 1:
        raise InvalidParamError(f"total_steps must be at least 1, got {total_steps}")
    if not 0 <= step < total_steps:
        raise InvalidParamError(f"step must be in 0..{total_steps - 1}, got {step}")
    if peak_lr <= 0:
        raise InvalidParamError(f"lr must be positive, got {peak_lr}")

    initial = peak_lr / div_factor
    final = initial / final_div_factor
    warmup = int(round(pct_start * total_steps))
    if step < warmup:
        p = step / warmup
        return initial + (peak_lr - initial) * (1.0 - math.cos(math.pi * p)) / 2.0
    p = (step - warmup) / (total_steps - warmup)
    return final + (peak_lr - final) * (1.0 + math.cos(math.pi * p)) / 2.0


def decay_param_groups(model: LexModel, weight_decay: float) -> list[dict[str, object]]:
    """Split parameters into the decayed matrices and everything else.

    Decayed: `qkv_w`, `proj_w`, `fc1_w`, `fc2_w`, and `head_w` when untied. NOT `embed`,
    NOT `pos`, NOT any bias, NOT LayerNorm gains — even though `embed` and `pos` are also
    2-D. Each group carries a `names` key so a test (and the UI) can read the split back.
    """
    decayed: list[tuple[str, torch.nn.Parameter]] = []
    plain: list[tuple[str, torch.nn.Parameter]] = []
    for name, param in model.named_parameters():
        if name.rsplit(".", 1)[-1] in DECAY_WEIGHTS:
            if param.ndim != 2:
                raise ComputeError(f"{name} is on the decay list but is not a matrix")
            decayed.append((name, param))
        else:
            plain.append((name, param))
    return [
        {
            "params": [p for _, p in decayed],
            "names": [n for n, _ in decayed],
            "weight_decay": float(weight_decay),
        },
        {
            "params": [p for _, p in plain],
            "names": [n for n, _ in plain],
            "weight_decay": 0.0,
        },
    ]


def batch_loss(model: LexModel, inputs: torch.Tensor, targets: torch.Tensor) -> torch.Tensor:
    """Mean next-token cross-entropy over the block, `ignore_index = PAD_ID`."""
    logits = model(inputs)
    return torch.nn.functional.cross_entropy(
        logits.reshape(-1, model.cfg.vocab_rows), targets.reshape(-1), ignore_index=PAD_ID
    )


def check_finite_loss(value: float, step: int) -> None:
    """Raise on a diverged run. Never clamp, never skip the step, never return a number."""
    if not math.isfinite(value):
        raise ComputeError(
            f"training diverged at step {step}: the loss is {value}. The learning rate is "
            "almost certainly too high for this configuration; lower it and re-run."
        )


@torch.no_grad()
def eval_loss(model: LexModel, split: np.ndarray, batch_size: int = 32) -> float:
    """Mean token cross-entropy (nats) over the fixed validation windows."""
    was_training = model.training
    model.eval()
    try:
        total, count = 0.0, 0
        for windows in iter_eval_windows(split, model.cfg.ctx, batch_size):
            targets = windows[:, 1:]
            n_tokens = int((targets != PAD_ID).sum())
            if n_tokens == 0:
                continue
            total += float(batch_loss(model, windows[:, :-1], targets)) * n_tokens
            count += n_tokens
    finally:
        model.train(was_training)
    if count == 0:
        raise ComputeError("eval_loss: the validation split holds no scoreable tokens")
    return total / count


@contextmanager
def _seeded_global_rng(seed: int):
    """Seed the global RNG (dropout masks read it) and restore it afterwards.

    Leaking a seeded global generator would make any later unseeded sampling depend on
    whether a training run happened first — the same trap the Geometry Lab documents.
    """
    previous = torch.get_rng_state()
    torch.manual_seed(int(seed))
    try:
        yield
    finally:
        torch.set_rng_state(previous)


def train_lex(
    *,
    stream: np.ndarray | Sequence[int],
    config: LexConfig | None = None,
    model: LexModel | None = None,
    steps: int = DEFAULT_STEPS,
    lr: float = DEFAULT_LR,
    batch_size: int = DEFAULT_BATCH,
    weight_decay: float = DEFAULT_WEIGHT_DECAY,
    seed: int = DEFAULT_SEED,
    val_fraction: float = VAL_FRACTION,
    progress_cb: ProgressCb | None = None,
    progress_every: int = 1,
) -> TrainResult:
    """Train a `LexModel` on a token stream and return the model plus its trajectory.

    Pass `config` to train from scratch, or `model` to continue training an existing one —
    that second form is the fine-tune path (FR-619), and because the caller owns the model
    object, its vocabulary travels with it instead of being silently re-derived.

    `progress_cb` receives a `TrainProgress` every `progress_every` steps and always on the
    final step, which is what the SSE route streams. A caller that also wants generated
    samples mid-run constructs the model itself and samples from it inside the callback.
    """
    if (config is None) == (model is None):
        raise InvalidParamError("exactly one of config/model must be provided")
    steps = int(steps)
    if not 1 <= steps <= MAX_STEPS:
        raise InvalidParamError(f"steps must be in 1..{MAX_STEPS}, got {steps}")
    batch_size = int(batch_size)
    if batch_size < 1:
        raise InvalidParamError(f"batch_size must be at least 1, got {batch_size}")
    if lr <= 0:
        raise InvalidParamError(f"lr must be positive, got {lr}")
    if weight_decay < 0:
        raise InvalidParamError(f"weight_decay must be non-negative, got {weight_decay}")
    if progress_every < 1:
        raise InvalidParamError(f"progress_every must be at least 1, got {progress_every}")

    stream = np.asarray(stream, dtype=np.int64)
    if model is None:
        assert config is not None
        model = LexModel(config, seed=seed)
    cfg = model.cfg
    if stream.size and (int(stream.max()) >= cfg.vocab_rows or int(stream.min()) < 0):
        raise InvalidParamError(
            f"token stream has ids outside 0..{cfg.vocab_rows - 1} — it was encoded with a "
            "different vocabulary than this model's"
        )
    train_split, val_split = split_stream(stream, cfg.ctx, val_fraction)

    groups = decay_param_groups(model, weight_decay)
    opt = torch.optim.AdamW(groups, lr=lr, betas=(0.9, 0.999), eps=1e-8)

    result = TrainResult(model=model, steps=steps, seed=int(seed))
    started = time.perf_counter()
    with _seeded_global_rng(seed):
        batch_gen = torch.Generator().manual_seed(int(seed) + 1)
        model.train()
        for step in range(steps):
            step_lr = onecycle_lr(step, steps, lr)
            for group in opt.param_groups:
                group["lr"] = step_lr

            inputs, targets = sample_batch(train_split, cfg.ctx, batch_size, batch_gen)
            loss = batch_loss(model, inputs, targets)
            loss_value = float(loss.detach())
            check_finite_loss(loss_value, step + 1)

            opt.zero_grad(set_to_none=True)
            loss.backward()
            torch.nn.utils.clip_grad_norm_(model.parameters(), GRAD_CLIP_NORM)
            opt.step()

            result.history.append({"step": step + 1, "loss": loss_value, "lr": step_lr})
            if step == 0:
                result.first_loss = loss_value
            result.final_loss = loss_value
            if progress_cb is not None and ((step + 1) % progress_every == 0 or step + 1 == steps):
                progress_cb(
                    TrainProgress(
                        step=step + 1,
                        total_steps=steps,
                        loss=loss_value,
                        lr=step_lr,
                        elapsed_s=time.perf_counter() - started,
                    )
                )
        model.eval()
        result.val_loss = eval_loss(model, val_split)
    result.elapsed_s = time.perf_counter() - started
    return result

File: code/frontend/src/lib/lexEngine/train.ts
/**
 * Real from-scratch training in the browser, to the recipe in `architecture.md` §
 * Training. Nothing here is simulated: every loss is a forward pass through
 * `LexModel`, every step an AdamW update of real gradients.
 *
 *   * loss        mean cross-entropy of `logits[:, :-1]` vs `x[:, 1:]`, ignore `<pad>`
 *   * optimizer   AdamW, betas (0.9, 0.999), eps 1e-8, DECOUPLED decay applied ONLY to
 *                 the 2-D weight matrices (qkv_w, proj_w, fc1_w, fc2_w, and head_w when
 *                 untied) — not embeddings, not positions, not biases, not LayerNorm
 *                 gains. This deliberately DIFFERS from the source, which decays every
 *                 parameter; we follow the standard convention and say so in the UI.
 *   * schedule    one-cycle with `lr` as the PEAK: initial = lr/25, final = initial/1e4,
 *                 pct_start = 0.3, cosine in both phases
 *   * clipping    global L2 norm 1.0, before the step
 *   * batching    contiguous 95/5 train/val split; each batch draws B start offsets
 *                 uniformly WITH REPLACEMENT from the training span; a window is ctx+1
 *                 tokens so the last input token has a target
 *
 * Determinism: one seeded RNG drives initialization, batch selection and dropout.
 * Bit-equality with PyTorch is **not** claimed and is not achievable — platform BLAS and
 * non-portable RNG streams make a browser run and a Python run two independent runs of
 * the same recipe. The golden test pins a fixed-weights forward/backward instead.
 */

import { computeError, invalidParam } from "../geoEngine/errors";
import {
  LexModel,
  backward,
  cloneWeights,
  crossEntropy,
  decayedWeightNames,
  initWeights,
  nonFiniteWeightNames,
  sfc32,
  weightNames,
  zeroGrads,
  type LexConfig,
  type WeightSet,
} from "./model";
import { EOS_ID, LexVocab, splitLines, tokenize } from "./vocab";

// --- constants (lex/config.py) -------------------------------------------------------

export const DEFAULT_STEPS = 400;
export const MAX_STEPS = 3000;
/** The PEAK of the one-cycle schedule, not a constant learning rate. */
export const DEFAULT_LR = 3e-3;
/**
 * 16, not 32 — see the note on `DEFAULT_BATCH` in `lex/config.py`, which this mirrors.
 * With `DEFAULT_CTX = 32` a default 400-step run is ~47 s of compute rather than ~193 s.
 * Held-out loss after 400 steps goes 2.294 -> 2.333; `steps = 500` recovers it (2.258)
 * at ~56 s for anyone who moves the slider.
 */
export const DEFAULT_BATCH = 16;
export const DEFAULT_WEIGHT_DECAY = 0.01;
export const GRAD_CLIP_NORM = 1.0;
export const ONECYCLE_PCT_START = 0.3;
export const ONECYCLE_DIV_FACTOR = 25.0;
export const ONECYCLE_FINAL_DIV_FACTOR = 1e4;
export const VAL_FRACTION = 0.05;
export const DEFAULT_SEED = 0;
/** Emit a sample every N steps while training so the user watches text improve. */
export const DEFAULT_SAMPLE_EVERY = 50;

export const ADAM_BETA1 = 0.9;
export const ADAM_BETA2 = 0.999;
export const ADAM_EPS = 1e-8;

// --- the learning-rate schedule ------------------------------------------------------

/** Python's `round`: half-to-even, so `w` matches the backend at every step count. */
export function pyRound(x: number): number {
  const f = Math.floor(x);
  const diff = x - f;
  if (diff > 0.5) return f + 1;
  if (diff < 0.5) return f;
  return f % 2 === 0 ? f : f + 1;
}

/**
 * One-cycle learning rate at 0-indexed step `i` of `steps`, with `peak` as the maximum.
 * Warms up cosine-wise from `peak/25` over the first 30% of the run, then anneals
 * cosine-wise to `peak/25/1e4`.
 */
export function oneCycleLr(i: number, steps: number, peak: number): number {
  const initial = peak / ONECYCLE_DIV_FACTOR;
  const final = initial / ONECYCLE_FINAL_DIV_FACTOR;
  const w = pyRound(ONECYCLE_PCT_START * steps);
  if (i < w) {
    const p = i / w;
    return initial + ((peak - initial) * (1 - Math.cos(Math.PI * p))) / 2;
  }
  const denom = steps - w;
  const p = denom > 0 ? (i - w) / denom : 1;
  return final + ((peak - final) * (1 + Math.cos(Math.PI * p))) / 2;
}

// --- the token stream ----------------------------------------------------------------

/**
 * Encode a corpus as ONE id stream, with `<eos>` closing every line that has a word.
 * The TypeScript port of `lex/train.py::token_stream`, and the contract's definition
 * (`architecture.md` § Training, "Token stream construction"):
 *
 *   > Split the text into lines. For each line with at least one word token, emit that
 *   > line's ids **followed by `<eos>`**. Blank lines emit nothing. Nothing is prepended.
 *
 * This is the ONLY way training data may be built. It used to be `vocab.encodeText`, a
 * flat tokenize with no `<eos>` at all, so the browser model never saw a line boundary
 * and could not learn verse — while the UI claimed both runtimes ran the same recipe.
 * On the shipped corpus at the `full` Dolch budget this yields exactly 19,071 ids of
 * which 3,071 are `<eos>`, the same numbers Python produces (pinned by a test).
 *
 * Nursery rhymes are line-shaped: the line, not the paragraph, is the unit the model can
 * plausibly learn to finish, which is also why generation renders `<eos>` as a line break.
 */
export function tokenStream(text: string, vocab: LexVocab): number[] {
  const ids: number[] = [];
  for (const line of splitLines(text)) {
    const lineIds = vocab.encode(tokenize(line));
    if (lineIds.length > 0) {
      for (const id of lineIds) ids.push(id);
      ids.push(EOS_ID);
    }
  }
  return ids;
}

// --- batching ------------------------------------------------------------------------

export interface TokenSplit {
  train: Int32Array;
  val: Int32Array;
}

/** Contiguous 95/5 split of the token stream. The val span is never trained on. */
export function splitTokens(tokens: ArrayLike<number>, valFraction = VAL_FRACTION): TokenSplit {
  const n = tokens.length;
  const nTrain = Math.floor(n * (1 - valFraction));
  return {
    train: Int32Array.from({ length: nTrain }, (_, i) => tokens[i]),
    val: Int32Array.from({ length: n - nTrain }, (_, i) => tokens[nTrain + i]),
  };
}

export interface Batch {
  /** (B*T) model inputs — the window's first `ctx` tokens. */
  ids: Int32Array;
  /** (B*T) targets — the window's last `ctx` tokens. */
  targets: Int32Array;
  B: number;
  T: number;
}

/**
 * `B` windows of `ctx+1` tokens, start offsets drawn uniformly WITH REPLACEMENT.
 * Sampling with replacement is the source's scheme and is what makes a "step" a fixed
 * unit of work independent of corpus length.
 */
export function sampleBatch(stream: Int32Array, ctx: number, B: number, rand: () => number): Batch {
  const span = ctx + 1;
  const maxStart = stream.length - span;
  if (maxStart < 0) {
    throw invalidParam(
      `the training text is too short: ${stream.length} tokens after tokenization, but one ` +
        `window at ctx = ${ctx} needs ${span}. Paste more text or lower ctx.`,
    );
  }
  const ids = new Int32Array(B * ctx);
  const targets = new Int32Array(B * ctx);
  for (let b = 0; b < B; b++) {
    const start = Math.min(maxStart, Math.floor(rand() * (maxStart + 1)));
    for (let t = 0; t < ctx; t++) {
      ids[b * ctx + t] = stream[start + t];
      targets[b * ctx + t] = stream[start + t + 1];
    }
  }
  return { ids, targets, B, T: ctx };
}

/** Deterministic non-overlapping windows — what validation and `evalLoss` walk. */
export function tilingBatches(stream: Int32Array, ctx: number, maxWindows = Infinity): Batch[] {
  const span = ctx + 1;
  const out: Batch[] = [];
  for (let s = 0; s + span <= stream.length && out.length < maxWindows; s += ctx) {
    const ids = new Int32Array(ctx);
    const targets = new Int32Array(ctx);
    for (let t = 0; t < ctx; t++) {
      ids[t] = stream[s + t];
      targets[t] = stream[s + t + 1];
    }
    out.push({ ids, targets, B: 1, T: ctx });
  }
  return out;
}

/** Mean token cross-entropy over deterministic windows, dropout off. */
export function evalLoss(model: LexModel, stream: Int32Array, maxWindows = 16): number {
  const batches = tilingBatches(stream, model.cfg.ctx, maxWindows);
  if (batches.length === 0) return NaN;
  let sum = 0;
  let count = 0;
  const V = model.cfg.vocabRows;
  for (const batch of batches) {
    const acts = model.forward(batch.ids, batch.B, batch.T, {});
    // crossEntropy divides by nValid; multiply back to accumulate a true token mean.
    const { ceSum, nValid } = crossEntropy(acts.logits, batch.targets, batch.B * batch.T, V);
    sum += ceSum;
    count += nValid;
  }
  return count > 0 ? sum / count : NaN;
}

// --- the optimizer -------------------------------------------------------------------

interface AdamState {
  m: Record<string, Float64Array>;
  v: Record<string, Float64Array>;
  t: number;
}

function newAdamState(cfg: LexConfig, ws: WeightSet): AdamState {
  const state: AdamState = { m: {}, v: {}, t: 0 };
  for (const name of weightNames(cfg)) {
    state.m[name] = new Float64Array(ws[name].length);
    state.v[name] = new Float64Array(ws[name].length);
  }
  return state;
}

// --- the run -------------------------------------------------------------------------

export interface TrainPoint {
  step: number;
  loss: number;
  lr: number;
  gradNorm: number;
  /** Present only on the steps validation ran. */
  valLoss?: number;
}

export type ProgressCb = (fraction: number, point: TrainPoint) => void;
export type SampleCb = (step: number, model: LexModel) => void;

export interface TrainOptions {
  cfg: LexConfig;
  /** The corpus, already encoded with the model's vocabulary. */
  tokens: ArrayLike<number>;
  steps?: number;
  lr?: number;
  batchSize?: number;
  weightDecay?: number;
  seed?: number;
  /** Continue from these weights instead of a fresh init (fine-tuning, FR-619). */
  initialWeights?: WeightSet;
  /** Validate + report every N steps (0 disables). */
  evalEvery?: number;
  sampleEvery?: number;
  onProgress?: ProgressCb;
  onSample?: SampleCb;
}

export interface TrainResult {
  weights: WeightSet;
  history: TrainPoint[];
  /** Mean CE on the held-out 5% after the last step. */
  valLoss: number;
  /** Mean CE on the training span before the first step — the "did it learn?" baseline. */
  initialTrainLoss: number;
  finalTrainLoss: number;
  steps: number;
  nTokens: number;
}

export function runTraining(opts: TrainOptions): TrainResult {
  const cfg = opts.cfg;
  const steps = Math.trunc(opts.steps ?? DEFAULT_STEPS);
  const peakLr = opts.lr ?? DEFAULT_LR;
  const batchSize = Math.trunc(opts.batchSize ?? DEFAULT_BATCH);
  const weightDecay = opts.weightDecay ?? DEFAULT_WEIGHT_DECAY;
  const seed = Math.trunc(opts.seed ?? DEFAULT_SEED);
  const evalEvery = Math.trunc(opts.evalEvery ?? Math.max(1, Math.floor(steps / 10)));
  const sampleEvery = Math.trunc(opts.sampleEvery ?? DEFAULT_SAMPLE_EVERY);

  if (!(steps >= 1 && steps <= MAX_STEPS)) throw invalidParam(`steps must be in 1..${MAX_STEPS}, got ${opts.steps}`);
  if (!(peakLr > 0 && Number.isFinite(peakLr))) throw invalidParam(`lr must be a finite number > 0, got ${opts.lr}`);
  if (!(batchSize >= 1)) throw invalidParam(`batch_size must be >= 1, got ${opts.batchSize}`);
  if (!(weightDecay >= 0)) throw invalidParam(`weight_decay must be >= 0, got ${opts.weightDecay}`);

  // A fine-tune whose base is already broken must fail HERE, naming the cause, rather
  // than a hundred lines later as "training diverged; try a lower learning rate".
  if (opts.initialWeights) {
    const bad = nonFiniteWeightNames(cfg, opts.initialWeights);
    if (bad.length > 0) {
      throw computeError(
        `cannot train from these weights: ${bad.join(", ")} contain non-finite values (NaN or ±Infinity), ` +
          "most likely from a weight edit that pushed them out of float32 range or from a diverged run. " +
          "Reset the edited tensor, or retrain from scratch.",
      );
    }
  }

  const { train, val } = splitTokens(opts.tokens);
  // Never mutate the caller's tensors: a fine-tune must leave its base model intact.
  const model = new LexModel(cfg, opts.initialWeights ? cloneWeights(opts.initialWeights) : initWeights(cfg, seed));
  const ws = model.weights;
  const names = weightNames(cfg);
  const decayed = new Set(decayedWeightNames(cfg));
  const adam = newAdamState(cfg, ws);
  const grads = zeroGrads(cfg);
  // A separate stream for data and dropout, so changing `steps` does not change the init.
  const rand = sfc32(seed + 1);

  const initialTrainLoss = evalLoss(model, train);
  const history: TrainPoint[] = [];
  const V = cfg.vocabRows;

  for (let step = 0; step < steps; step++) {
    const batch = sampleBatch(train, cfg.ctx, batchSize, rand);
    for (const name of names) grads[name].fill(0);

    const acts = model.forward(batch.ids, batch.B, batch.T, { dropout: cfg.dropout, rand });
    const { ceSum, nValid, dLogits } = crossEntropy(acts.logits, batch.targets, batch.B * batch.T, V);
    const loss = ceSum / nValid;
    if (!Number.isFinite(loss)) {
      throw computeError(`training diverged at step ${step + 1} (non-finite loss); try a lower learning rate`);
    }
    backward(model, acts, dLogits, grads);

    // Global-norm clipping at 1.0, before the step (torch clip_grad_norm_ semantics).
    let sq = 0;
    for (const name of names) {
      const g = grads[name];
      for (let i = 0; i < g.length; i++) sq += g[i] * g[i];
    }
    const gradNorm = Math.sqrt(sq);
    const clip = Math.min(1, GRAD_CLIP_NORM / (gradNorm + 1e-6));

    const lr = oneCycleLr(step, steps, peakLr);
    adam.t++;
    const bc1 = 1 - Math.pow(ADAM_BETA1, adam.t);
    const bc2 = 1 - Math.pow(ADAM_BETA2, adam.t);
    for (const name of names) {
      const p = ws[name];
      const g = grads[name];
      const m = adam.m[name];
      const v = adam.v[name];
      const decay = decayed.has(name) ? lr * weightDecay : 0;
      for (let i = 0; i < p.length; i++) {
        const gi = g[i] * clip;
        m[i] = ADAM_BETA1 * m[i] + (1 - ADAM_BETA1) * gi;
        v[i] = ADAM_BETA2 * v[i] + (1 - ADAM_BETA2) * gi * gi;
        // AdamW: decoupled decay first, then the adaptive step.
        const decayed_ = decay > 0 ? p[i] - decay * p[i] : p[i];
        p[i] = Math.fround(decayed_ - (lr * (m[i] / bc1)) / (Math.sqrt(v[i] / bc2) + ADAM_EPS));
      }
    }

    const point: TrainPoint = { step: step + 1, loss, lr, gradNorm };
    if (evalEvery > 0 && ((step + 1) % evalEvery === 0 || step === steps - 1) && val.length > cfg.ctx) {
      point.valLoss = evalLoss(model, val);
    }
    history.push(point);
    opts.onProgress?.((step + 1) / steps, point);
    if (sampleEvery > 0 && (step + 1) % sampleEvery === 0) opts.onSample?.(step + 1, model);
  }

  const finalTrainLoss = evalLoss(model, train);
  const valLoss = val.length > cfg.ctx ? evalLoss(model, val) : NaN;
  if (!Number.isFinite(finalTrainLoss)) throw computeError("training produced a non-finite loss; refusing to save");

  return {
    weights: ws,
    history,
    valLoss,
    initialTrainLoss,
    finalTrainLoss,
    steps,
    nTokens: opts.tokens.length,
  };
}

/** Guard against a corpus too small for even one window, before any work is done. */
export function assertTrainable(nTokens: number, ctx: number): void {
  const nTrain = Math.floor(nTokens * (1 - VAL_FRACTION));
  if (nTrain < ctx + 1) {
    throw invalidParam(
      `this text has ${nTokens} in-budget tokens, of which ${nTrain} are the training ` +
        `split — one window at ctx = ${ctx} needs ${ctx + 1}. Paste more text or lower ctx.`,
    );
  }
}
